### PR TITLE
[LoopVectorize][AArch64] Add limited support for scalable vectorisation of i1 types

### PR DESCRIPTION
**Pull request by david-arm**
Previously isElementTypeLegalForScalableVector returned false for i1 types, which also prevented vectorisation of loops with i1 reductions. This is overkill - we only need to disable vectorisation for loads and/or stores of i1 types. I've added i1 as a legal type, but changed the cost model to return an invalid cost for loads and stores.

### CODE REVIEW

**Review by paulwalker-arm — COMMENTED** [review]
In general the patch looks good to me but we do support, and generate (see sve-extract-vector-to-predicate-store.ll), normal loads and stores of `<vscale x 16 x i1>` types.  Do you know what the cost model originally reported for these?

 If they already return an invalid cost then this patch makes nothing worse and so you're good to go. However, if that's not the case then `AArch64TTIImpl::getMemoryOpCost()` should be updated to allow for that one predicate type.

I'll add, I do see examples of longer i1 vector types (e.g. `<vscale x 32 x i1>` but I don't know if they're legitimate cases.  For the ACLE we plant `i8` based accesses so I wouldn't expect to see anything other than `<vscale x 16 x i1>` coming from ACLE code. I suppose if you do add support then it'll be easy enough to support all multiples of 16.

**Comment by david-arm** [comment]
> If they already return an invalid cost then this patch makes nothing worse and so you're good to go. However, if that's not the case then `AArch64TTIImpl::getMemoryOpCost()` should be updated to allow for that one predicate type.

Hi @paulwalker-arm, good suggestion. I've added a new commit that supports valid costs for multiples of <vscale x 16 x i1>. Enjoy!

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp [hunk 1]
```diff
@@ -3259,8 +3265,12 @@ InstructionCost AArch64TTIImpl::getMemoryOpCost(unsigned Opcode, Type *Ty,
   // of <vscale x 1 x eltty> yet, so return an invalid cost to avoid selecting
   // it. This change will be removed when code-generation for these types is
   // sufficiently reliable.
+  // We also cannot handle loads or stores involving scalable vectors of i1.
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:3268]: Perhaps "We also only support full register predicate loads and stores."?

**Review by paulwalker-arm — APPROVED** [review]

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp (hunk 1 above)
Comment (david-arm) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp]: Done!

**Pull request merged**